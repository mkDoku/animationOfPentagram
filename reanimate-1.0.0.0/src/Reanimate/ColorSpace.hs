{-|
Copyright   : Written by David Himmelstrup
License     : Unlicense
Maintainer  : lemmih@gmail.com
Stability   : experimental
Portability : POSIX
-}
module Reanimate.ColorSpace
  ( Nanometer
  , coneSensitivity
  , bigXYZCoordinates
  , lightXYZCoordinates
  , nmToColor
  ) where

import qualified Data.ByteString.Lazy       as BS
import           Data.Colour.CIE
import           Data.Csv
import           Data.Map                   (Map)
import qualified Data.Map                   as Map
import           Data.Maybe
import qualified Data.Vector                as V
import           Paths_reanimate
import           System.IO.Unsafe

-- | Wavelengths in nanometers.
type Nanometer = Integer

{-# NOINLINE lightXYZCoordinates #-}
-- | (small) xyz values for each wavelength of light.
lightXYZCoordinates :: Map Nanometer (Double, Double, Double)
lightXYZCoordinates = unsafePerformIO $ do
  dat <- BS.readFile =<< getDataFileName "data/CIExyz.csv"
  case decode NoHeader dat of
    Left err -> error err
    Right vec -> return $ Map.fromList
      [ (nm, (x,y,z)) | (nm,x,y,z) <- V.toList vec, nm <= 700 ]

{-# NOINLINE bigXYZCoordinates #-}
-- | (big) XYZ values for each wavelength of light.
bigXYZCoordinates :: Map Nanometer (Double, Double, Double)
bigXYZCoordinates = unsafePerformIO $ do
  dat <- BS.readFile =<< getDataFileName "data/CIE_XYZ.csv"
  case decode NoHeader dat of
    Left err -> error err
    Right vec -> return $ Map.fromList
      [ (nm, (x,y,z)) | (nm,x,y,z) <- V.toList vec, nm <= 700 ]

-- | Helper function for converting a wavelength of light into the
--   perceived color.
nmToColor :: Nanometer -> Maybe (Colour Double)
nmToColor nm = do
  (x, y, z) <- Map.lookup nm bigXYZCoordinates
  return $ cieXYZ x y z

{-# NOINLINE coneSensitivity #-}
-- (Long, Medium, Short)
-- | Cone sensitivity by light wavelength.
coneSensitivity :: Map Nanometer (Double, Double, Double)
coneSensitivity = unsafePerformIO $ do
  dat <- BS.readFile =<< getDataFileName "data/cone_sensitivity_lms.csv"
  case decode NoHeader dat of
    Left err -> error err
    Right vec -> return $ Map.fromList
      [ (nm, (l,m,fromMaybe 0 s)) | (nm,l,m,s) <- V.toList vec, nm <= 700 ]
