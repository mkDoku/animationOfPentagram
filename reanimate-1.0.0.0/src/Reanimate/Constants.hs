{- |
  Reanimate configures a consistent, default canvas. The values of this default
  can be observed via the constants in this module. Keep in mind, these values
  describe the /default/ canvas and will not apply to custom viewports.
-}
module Reanimate.Constants
  ( screenWidth
  , screenHeight
  , screenTop
  , screenBottom
  , screenLeft
  , screenRight
  , defaultDPI
  , defaultStrokeWidth
  ) where

import           Graphics.SvgTree

-- | Number of units from the left-most point to the right-most point on the screen.
screenWidth :: Fractional a => a

-- | Number of units from the bottom to the top of the screen.
screenHeight :: Fractional a => a

-- | Position of the top of the screen.
screenTop :: Fractional a => a

-- | Position of the bottom of the screen.
screenBottom :: Fractional a => a

-- | Position of the left side of the screen.
screenLeft :: Fractional a => a

-- | Position of the right side of the screen.
screenRight :: Fractional a => a

screenWidth = 16
screenHeight = 9
screenTop = screenHeight/2
screenBottom = -screenHeight/2
screenLeft = -screenWidth/2
screenRight = screenWidth/2

-- | SVG allows measurements in inches which have to be converted to local units.
--   This value describes how many local units there are in an inch.
defaultDPI :: Dpi
defaultDPI = 96

-- | Default thickness of lines.
defaultStrokeWidth :: Double
defaultStrokeWidth = 0.05
