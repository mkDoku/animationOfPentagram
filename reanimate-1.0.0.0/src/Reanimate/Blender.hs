{-# LANGUAGE OverloadedStrings #-}
{-|
  [Blender](https://www.blender.org/) is a free and open-source 3D graphics toolkit.
  It is usually used through a graphical user-interface but can also be
  scripted via Python. These Blender scripts can access 100% of Blender's
  functionality and offer a convenient way of coding 3D effects.

  Running Blender can be time-consuming but heavy caching means scripts
  are only run when they change.

  Blender cheatsheet:

> # To generate with a transparent background, set film_transparent = True:
> bpy.context.scene.render.film_transparent = True
>
> # Filmic is great for photorealism but bad for animations.
> # If you want your textures to keep their exact color values,
> # set the view_transform to 'Standard':
> bpy.context.scene.view_settings.view_transform = 'Standard'
>
> # Blender's default render engine is 'EEVEE', fast but not a raytracer.
> # To switch to raytracing, set the engine to 'CYCLES':
> bpy.context.scene.render.engine = 'CYCLES'
>
> # Rendering at full resolution can be slow. When developing, try
> # decreasing the resolution_percentage for faster renders.
> bpy.context.scene.render.resolution_percentage = 10
>
> # The resolution of the final image are set by resolution_x and resolution_y:
> bpy.context.scene.render.resolution_x = 320
> bpy.context.scene.render.resolution_y = 180

-}
module Reanimate.Blender
  ( blender
  , blender'
  ) where

import           Data.Hashable
import           Data.Text                  (Text)
import qualified Data.Text.IO               as T
import           Graphics.SvgTree           (Tree)
import           Reanimate.Animation
import           Reanimate.Cache
import           Reanimate.Constants
import           Reanimate.Misc
import           Reanimate.Parameters
import           Reanimate.Raster
import           Reanimate.Svg.Constructors
import           System.FilePath            (replaceExtension, (<.>))
import           System.IO.Unsafe           (unsafePerformIO)

-- | Run a Blender script and embed the resulting image file. The
--   image will be scaled to fit the screen exactly (assuming a default
--   canvas layout). Note that Blender resolution defaults to 1920x1080
--   but can be changed in the script code.
blender :: Text -> SVG
blender script =
  unsafePerformIO $ mkBlenderImage script

-- | Generate Blender image as a separate PNG file. Can be embedded with
--   `mkImage`.
blender' :: Text -> FilePath
blender' script =
  unsafePerformIO $ mkBlenderImage' script

mkBlenderImage :: Text -> IO Tree
mkBlenderImage script | pNoExternals = pure $ mkText script
mkBlenderImage script =
  mkImage screenWidth screenHeight <$> mkBlenderImage' script

mkBlenderImage' :: Text -> IO FilePath
mkBlenderImage' _ | pNoExternals = pure "/blender/has/been/disabled"
mkBlenderImage' script = cacheFile template $ \target -> do
    exec <- requireExecutable "blender"
    let py_file = replaceExtension target "py"
    T.writeFile py_file script
    runCmd exec [ "--background","--render-format", "PNG"
                , "--python-exit-code", "1"
                , "--render-output", target, "--python", py_file]
  where
    template = encodeInt (hash script) <.> "png"
