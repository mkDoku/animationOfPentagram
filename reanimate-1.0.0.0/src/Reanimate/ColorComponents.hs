{-# LANGUAGE RecordWildCards #-}
{- |
  Colors are three dimensional and can be projected into many color spaces
  with different properties.

  Interpolating directly in the RGB color space is unintuitive and rarely useful.
  If you want to transition through color, you most likely want either the XYZ space
  (for physically accurate color transitions) or the LAB space (for esthetically
  pleasing colors).
-}
module Reanimate.ColorComponents
  ( ColorComponents(..)
  , rgbComponents
  , hsvComponents
  , labComponents
  , xyzComponents
  , lchComponents
  , interpolate
  , interpolateRGB8
  , interpolateRGBA8
  , toRGB8
  , fromRGB8
  ) where

import           Codec.Picture
import           Codec.Picture.Types
import           Data.Colour
import           Data.Colour.CIE
import           Data.Colour.CIE.Illuminant (d65)
import           Data.Colour.RGBSpace
import           Data.Colour.RGBSpace.HSV
import           Data.Colour.SRGB
import           Data.Fixed
import           Reanimate.Ease

-- | Constructor and destructor for color's three components.
data ColorComponents = ColorComponents
  { colorUnpack :: Colour Double -> (Double, Double, Double)
    -- ^ Unpack a color into its three components.
  , colorPack   :: Double -> Double -> Double -> Colour Double
    -- ^ Restore a color from three coordinates.
  }

-- | > interpolate rgbComponents yellow blue
--
--   <<docs/gifs/doc_rgbComponents.gif>>
rgbComponents :: ColorComponents
rgbComponents = ColorComponents rgbUnpack sRGB
  where
    rgbUnpack :: Colour Double -> (Double, Double, Double)
    rgbUnpack c =
      case toSRGB c of
        RGB r g b -> (r,g,b)

-- | > interpolate hsvComponents yellow blue
--
--   <<docs/gifs/doc_hsvComponents.gif>>
hsvComponents :: ColorComponents
hsvComponents = ColorComponents unpack pack
  where
    unpack = hsvView.toSRGB
    pack a b c = uncurryRGB sRGB $ hsv a b c

-- | > interpolate labComponents yellow blue
--
--   <<docs/gifs/doc_labComponents.gif>>
labComponents :: ColorComponents
labComponents = ColorComponents unpack pack
  where
    unpack = cieLABView d65
    pack = cieLAB d65

-- | > interpolate xyzComponents yellow blue
--
--   <<docs/gifs/doc_xyzComponents.gif>>
xyzComponents :: ColorComponents
xyzComponents = ColorComponents cieXYZView cieXYZ

-- | > interpolate lchComponents yellow blue
--
--   <<docs/gifs/doc_lchComponents.gif>>
lchComponents :: ColorComponents
lchComponents = ColorComponents unpack pack
  where
    toDeg,toRad :: Double -> Double
    toRad deg = deg/180 * pi
    toDeg rad = rad/pi * 180
    unpack :: Colour Double -> (Double, Double, Double)
    unpack color =
      let (l,a,b) = cieLABView d65 color
          c = sqrt (a*a + b*b)
          h :: Double
          h = (toDeg(atan2 b a) + 360) `mod'` 360
          isZero = round (c*10000) == (0::Integer)
      in (l, c, if isZero then 0/0 else h)
    pack l c h =
      cieLAB d65 l (cos (toRad h) * c) (sin (toRad h) * c)

-- | Smoothly interpolate between two colors using the given color components.
interpolate :: ColorComponents -> Colour Double -> Colour Double -> (Double -> Colour Double)
interpolate ColorComponents{..} from to = \d ->
    colorPack (a1 + (a2-a1)*d) (b1 + (b2-b1)*d) (c1 + (c2-c1)*d)
  where
    (a1,b1,c1) = colorUnpack from
    (a2,b2,c2) = colorUnpack to

-- | Convenience interpolation function for RGB8 values.
interpolateRGB8 :: ColorComponents -> PixelRGB8 -> PixelRGB8 -> (Double -> PixelRGB8)
interpolateRGB8 comps from to = toRGB8 . interpolate comps (fromRGB8 from) (fromRGB8 to)

-- | Convenience interpolation function for RGBA8 values.
interpolateRGBA8 :: ColorComponents -> PixelRGBA8 -> PixelRGBA8 -> (Double -> PixelRGBA8)
interpolateRGBA8 comps from to = \t ->
  case interp t of
    PixelRGB8 r g b ->
      let alpha = fromToS (fromIntegral $ pixelOpacity from) (fromIntegral $ pixelOpacity to) t
      in PixelRGBA8 r g b (round alpha)
  where
    interp = interpolateRGB8 comps (dropTransparency from) (dropTransparency to)

-- | Convenience function for expressing a color as an RGB8 value.
toRGB8 :: Colour Double -> PixelRGB8
toRGB8 c = PixelRGB8 r g b
  where
    RGB r g b = toSRGBBounded c

-- | Convenience function for expressing an RGB8 value as a color.
fromRGB8 :: PixelRGB8 -> Colour Double
fromRGB8 (PixelRGB8 r g b) = sRGB24 r g b
