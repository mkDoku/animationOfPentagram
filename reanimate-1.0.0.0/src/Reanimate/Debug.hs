{-|
Copyright   : Written by David Himmelstrup
License     : Unlicense
Maintainer  : lemmih@gmail.com
Stability   : experimental
Portability : POSIX
-}
module Reanimate.Debug
  ( traceSVG
  , traceA
  , playTraces
  )
where

import           Data.IORef
import           System.IO.Unsafe
import           Control.Exception
import           Reanimate.Animation
import           Reanimate.Svg
import           Reanimate.LaTeX
import           Reanimate.Constants
import           Text.Printf
import qualified Data.Text                     as T

{-# NOINLINE traceBuffer #-}
traceBuffer :: IORef [Animation]
traceBuffer = unsafePerformIO (newIORef [])

{-# NOINLINE traceSVG #-}
-- | Add SVG image to trace stack.
traceSVG :: SVG -> a -> a
traceSVG = traceA . staticFrame (recip 60)

{-# NOINLINE traceA #-}
-- | Add animation to trace stack.
traceA :: Animation -> a -> a
traceA a v = unsafePerformIO $ do
  modifyIORef' traceBuffer (a :)
  evaluate v

{-# NOINLINE playTraces #-}
-- | Evaluate argument and play back the trace stack.
playTraces :: a -> Animation
playTraces v = unsafePerformIO $ do
  _   <- evaluate v
  lst <- atomicModifyIORef' traceBuffer (\x -> ([], reverse x))
  let n = length lst :: Int
  return $ foldr
    seqA
    (pause 0)
    [ f `parA` staticFrame (duration f) (counter i n) | (i, f) <- zip [1 :: Int ..] lst ]
 where
  counter a b = mkGroup
    [ withStrokeWidth defaultStrokeWidth
    $ withStrokeColor "black"
    $ translate 6.5 4
    $ center
    $ latex
    $ T.pack
    $ printf "%d/%d" a b
    , withStrokeWidth 0
    $ withFillColor "white"
    $ translate 6.5 4
    $ center
    $ latex
    $ T.pack
    $ printf "%d/%d" a b
    ]
