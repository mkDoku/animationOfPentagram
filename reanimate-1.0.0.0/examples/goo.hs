#!/usr/bin/env stack
-- stack runghc --package reanimate
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where

import           Control.Lens

import           Graphics.SvgTree
import           Reanimate

main :: IO ()
main = reanimate $ playThenReverseA $ mkAnimation 5 $ \t ->
    let s = fromToS 0 1.5 $ curveS 2 t in
    mkGroup
    [ mkBackground "black"
    , FilterTree $ mkFilter "blur"
      [FEGaussianBlur $ defaultSvg
        & gaussianBlurStdDeviationX .~ Num dev
        & filterResult ?~ "blur"
      ] & filterWidth .~ pure (Percent 3)
        & filterX .~ pure (Percent (-1))
        & filterHeight .~ pure (Percent 3)
        & filterY .~ pure (Percent (-1))
    , FilterTree $ mkFilter "goo"
      [FEGaussianBlur $ defaultSvg
        & gaussianBlurStdDeviationX .~ Num dev
        & filterResult ?~ "blur"
      ,FEColorMatrix $ defaultSvg
        & colorMatrixType .~ Matrix
        & colorMatrixValues .~ "1 0 0 0 0 \
                               \0 1 0 0 0 \
                               \0 0 1 0 0 \
                               \0 0 0 " ++ show (sharpness*2) ++ " -" ++ show sharpness
        & filterResult .~ pure "goo"
      ,FEComposite $ defaultSvg
        & compositeIn .~ pure SourceGraphic
        & compositeIn2 .~ pure (SourceRef "goo")
        & compositeOperator .~ CompositeAtop
      ] & filterWidth .~ pure (Percent 3)
        & filterX .~ pure (Percent (-1))
        & filterHeight .~ pure (Percent 3)
        & filterY .~ pure (Percent (-1))
    , translate 0 (radius*2.2) $ withFillColor "red" $ mkGroup
      [ translate (s*(-radius)) 0 circ
      , translate (s*radius) 0 circ
      ]
    , withFillColor "red" $ mkGroup
      [ translate (s*(-radius)) 0 circ
      , translate (s*radius) 0 circ
      ] & filterRef .~ pure (Ref "blur")
    , translate 0 (-radius*2.2) $ withFillColor "red" $ mkGroup
      [ translate (s*(-radius)) 0 circ
      , translate (s*radius) 0 circ
      ] & filterRef .~ pure (Ref "goo")
    ]
  where
    sharpness = 10 :: Integer
    dev = 0.2
    radius = 1
    circ = mkCircle radius

mkFilter :: String -> [FilterElement] -> Filter
mkFilter ident fe = defaultSvg & filterChildren .~ fe & attrId ?~ ident
