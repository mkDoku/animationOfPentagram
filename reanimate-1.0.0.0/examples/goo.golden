0<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(0, 0)"><circle r="1" /></g><g transform="translate(0, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(0, 0)"><circle r="1" /></g><g transform="translate(0, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(0, 0)"><circle r="1" /></g><g transform="translate(0, 0)"><circle r="1" /></g></g></g></g></g></svg>
1<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-0.148148, 0)"><circle r="1" /></g><g transform="translate(0.148148, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-0.148148, 0)"><circle r="1" /></g><g transform="translate(0.148148, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-0.148148, 0)"><circle r="1" /></g><g transform="translate(0.148148, 0)"><circle r="1" /></g></g></g></g></g></svg>
2<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-0.592593, 0)"><circle r="1" /></g><g transform="translate(0.592593, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-0.592593, 0)"><circle r="1" /></g><g transform="translate(0.592593, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-0.592593, 0)"><circle r="1" /></g><g transform="translate(0.592593, 0)"><circle r="1" /></g></g></g></g></g></svg>
3<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-1.166667, 0)"><circle r="1" /></g><g transform="translate(1.166667, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-1.166667, 0)"><circle r="1" /></g><g transform="translate(1.166667, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-1.166667, 0)"><circle r="1" /></g><g transform="translate(1.166667, 0)"><circle r="1" /></g></g></g></g></g></svg>
4<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-1.462963, 0)"><circle r="1" /></g><g transform="translate(1.462963, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-1.462963, 0)"><circle r="1" /></g><g transform="translate(1.462963, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-1.462963, 0)"><circle r="1" /></g><g transform="translate(1.462963, 0)"><circle r="1" /></g></g></g></g></g></svg>
5<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-1.462963, 0)"><circle r="1" /></g><g transform="translate(1.462963, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-1.462963, 0)"><circle r="1" /></g><g transform="translate(1.462963, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-1.462963, 0)"><circle r="1" /></g><g transform="translate(1.462963, 0)"><circle r="1" /></g></g></g></g></g></svg>
6<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-1.166667, 0)"><circle r="1" /></g><g transform="translate(1.166667, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-1.166667, 0)"><circle r="1" /></g><g transform="translate(1.166667, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-1.166667, 0)"><circle r="1" /></g><g transform="translate(1.166667, 0)"><circle r="1" /></g></g></g></g></g></svg>
7<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-0.592593, 0)"><circle r="1" /></g><g transform="translate(0.592593, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-0.592593, 0)"><circle r="1" /></g><g transform="translate(0.592593, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-0.592593, 0)"><circle r="1" /></g><g transform="translate(0.592593, 0)"><circle r="1" /></g></g></g></g></g></svg>
8<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(-0.148148, 0)"><circle r="1" /></g><g transform="translate(0.148148, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(-0.148148, 0)"><circle r="1" /></g><g transform="translate(0.148148, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(-0.148148, 0)"><circle r="1" /></g><g transform="translate(0.148148, 0)"><circle r="1" /></g></g></g></g></g></svg>
9<svg viewBox="-8 -4.5 16 9" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="none"><g stroke-width="0.05" transform="scale(1, -1)"><g><g stroke-width="0" fill="#000000" transform="translate(-8, -4.5)" fill-opacity="1"><rect width="16" height="9" /></g><filter id="blur" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" /></filter><filter id="goo" width="300%" height="300%" x="-100%" y="-100%">  <feGaussianBlur result="blur" stdDeviation="0.2" />  <feColorMatrix result="goo" values="1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 20 -10" />  <feComposite in="SourceGraphic" in2="goo" operator="atop" /></filter><g transform="translate(0, 2.2)"><g fill="#FF0000"><g transform="translate(0, 0)"><circle r="1" /></g><g transform="translate(0, 0)"><circle r="1" /></g></g></g><g fill="#FF0000" filter="url(#blur)"><g transform="translate(0, 0)"><circle r="1" /></g><g transform="translate(0, 0)"><circle r="1" /></g></g><g transform="translate(0, -2.2)"><g fill="#FF0000" filter="url(#goo)"><g transform="translate(0, 0)"><circle r="1" /></g><g transform="translate(0, 0)"><circle r="1" /></g></g></g></g></g></svg>
